File: SRAudioKit/Sources/SRAUGraph.m
//
//  SRAUGraph.m
//  SRAudioKit
//
//  Created by Heeseung Seo on 2015. 9. 18..
//  Copyright © 2015년 Seorenn. All rights reserved.
//

#import "SRAUGraph.h"
#import "SRAudioKitUtils.h"
#import "SRAUNode.h"
#import "SRAudioUnit.h"

#import "TargetConditionals.h"

@import CoreAudio;
@import AudioToolbox;

// Callback Definitions
static OSStatus SRAUGraphNodeInputCallback(void                          *inRefCon,
                                           AudioUnitRenderActionFlags    *ioActionFlags,
                                           const AudioTimeStamp          *inTimeStamp,
                                           UInt32                        inBusNumber,
                                           UInt32                        inNumberFrames,
                                           AudioBufferList               *ioData);

@interface SRAUGraph () {
    AUGraph _graph;
}

@end

@implementation SRAUGraph

#pragma mark - Life Cycle

- (nonnull instancetype)init {
    self = [super init];
    if (self) {
        CheckOSStatus(NewAUGraph(&_graph), @"NewAUGraph");
    }
    return self;
}

- (void)dealloc {
    [self dispose];
}

#pragma mark - Properties Getters 



#pragma mark - AUGraph Wrappers

- (BOOL)initialize {
    return CheckOSStatus(AUGraphInitialize(_graph), @"AUGraphInitialize");
}

- (BOOL)uninitialize {
    return CheckOSStatus(AUGraphUninitialize(_graph), @"AUGraphUninitialize");
}


- (nullable SRAUNode *)addNodeWithAudioComponentDescription:(AudioComponentDescription)audioComponentDescription {
    AUNode node;
    if (CheckOSStatus(AUGraphAddNode(_graph, &audioComponentDescription, &node), @"AUGraphAddNode") == NO) {
        return nil;
    };
    
    /*
    SRAudioUnit *audioUnit = [self nodeInfo:node];
    if (audioUnit == nil) {
        return nil;
    }
     */

    SRAUNode *nodeObject = [[SRAUNode alloc] initWithNode:node audioComponentDescription:audioComponentDescription];
    
    return nodeObject;
}

- (nullable SRAudioUnit *)nodeInfo:(nonnull SRAUNode *)node withAudioComponentDescription:(AudioComponentDescription)audioComponentDescription {
    AudioComponentDescription description;
    AudioUnit audioUnit;
    
    if (CheckOSStatus(AUGraphNodeInfo(_graph, node.node, &description, &audioUnit), @"AUGraphNodeInfo") == NO) {
        return nil;
    }
    
    SRAudioUnit *au = [[SRAudioUnit alloc] initWithAudioUnit:audioUnit];
    return au;
}

- (BOOL)setNodeDelegate:(nonnull id<SRAUNodeDelegate>)delegate destInputNumber:(UInt32)destInputNumber forNode:(nonnull SRAUNode *)node {
    AURenderCallbackStruct callbackStruct;
    callbackStruct.inputProc = SRAUGraphNodeInputCallback;
    callbackStruct.inputProcRefCon = (__bridge void *)node;

    BOOL res = CheckOSStatus(AUGraphSetNodeInputCallback(_graph, node.node, destInputNumber, &callbackStruct), @"AUGraphSetNodeInputCallback");
    if (res) {
        node.delegate = delegate;
    }
    return res;
}

- (BOOL)connectSourceNodeInput:(nonnull SRAUNode *)sourceNode sourceOutputNumber:(UInt32)sourceOutputNumber destNode:(nonnull SRAUNode *)destNode destOutputNumber:(UInt32)destInputNumber {
    return CheckOSStatus(AUGraphConnectNodeInput(_graph, sourceNode.node, sourceOutputNumber, destNode.node, destInputNumber), @"AUGraphConnectNodeInput");
}

- (BOOL)disconnectNodeInput:(nonnull SRAUNode *)destNode destInputNumber:(UInt32)destInputNumber {
    return CheckOSStatus(AUGraphDisconnectNodeInput(_graph, destNode.node, destInputNumber), @"AUGraphDisconnectNodeInput");
}

- (BOOL)removeNode:(nonnull SRAUNode *)node {
    return CheckOSStatus(AUGraphRemoveNode(_graph, node.node), @"AUGraphRemoveNode");
}

- (BOOL)update:(BOOL)synchronize {
    // TODO: This method returns incorrect values by some cases...
    if (synchronize) {
        return CheckOSStatus(AUGraphUpdate(_graph, NULL), @"AUGraphUpdate");
    }
    else {
        Boolean value = false;
        if (CheckOSStatus(AUGraphUpdate(_graph, &value), @"AUGraphUpdate") == NO) {
            return NO;
        }
        
        return value;
    }
    return NO;
}

- (BOOL)open {
    return CheckOSStatus(AUGraphOpen(_graph), @"AUGraphOpen");
}

- (BOOL)start {
    return CheckOSStatus(AUGraphStart(_graph), @"AUGraphStart");
}

- (BOOL)stop {
    return CheckOSStatus(AUGraphStop(_graph), @"AUGraphStop");
}

- (BOOL)close {
    return CheckOSStatus(AUGraphClose(_graph), @"AUGraphClose");
}

- (BOOL)dispose {
    return CheckOSStatus(DisposeAUGraph(_graph), @"DispostAUGraph");
}

@end

static OSStatus SRAUGraphNodeInputCallback(void                          *inRefCon,
                                           AudioUnitRenderActionFlags    *ioActionFlags,
                                           const AudioTimeStamp          *inTimeStamp,
                                           UInt32                        inBusNumber,
                                           UInt32                        inNumberFrames,
                                           AudioBufferList               *ioData) {
    SRAUNode *node = (__bridge SRAUNode *)inRefCon;
    OSStatus error = noErr;
    
    error = AudioUnitRender(node.audioUnit.audioUnit, ioActionFlags, inTimeStamp, inBusNumber, inNumberFrames, node.audioUnit.audioBufferList);
    
    if (CheckOSStatus(error, @"AudioUnitRender") == NO) {
        return error;
    }
    
    if (node.delegate == nil) return error;
    
    [node.delegate auNode:node
        didTakeBufferList:node.audioUnit.audioBufferList
               bufferSize:inNumberFrames
         numberOfChannels:node.audioUnit.audioStreamBasicDescription.mChannelsPerFrame];
    
    return error;
}

File: SRAudioKit/Sources/SRAUGraph.swift
//
//  SRAUGraph.swift
//  SRAudioKit
//
//  Created by Heeseung Seo on 2015. 9. 24..
//  Copyright © 2015년 Seorenn. All rights reserved.
//

import Foundation
import CoreAudio
import AudioToolbox
import SRAudioKitPrivates

open class SRAUGraph {
  let graph: AUGraph
  
  public init() {
    var graphInstance: AUGraph? = nil  //= AUGraph()
    NewAUGraph(&graphInstance)
    self.graph = graphInstance!
  }
  
  deinit {
    DisposeAUGraph(self.graph)
  }
  
  open var running: Bool {
    var value = DarwinBoolean(false)
    let res = AUGraphIsRunning(self.graph, &value)
    guard res == noErr else { return false }
    
    return value.boolValue
  }
  
  open func clearConnections() throws {
    let res = AUGraphClearConnections(self.graph)
    guard res == noErr else {
      throw SRAudioError.osStatusError(status: res, description: "SRAUGraph.clearConnections()")
    }
  }
  
  open func addNode(_ componentDescription: AudioComponentDescription) throws -> SRAUNode {
    var node = AUNode()
    var desc = componentDescription
    let res = AUGraphAddNode(self.graph, &desc, &node)
    guard res == noErr else {
      throw SRAudioError.osStatusError(status: res, description: "SRAUGraph.addNode()")
    }
    
    return SRAUNode(node: node)
  }
  
  open func open() throws {
    let res = AUGraphOpen(self.graph)
    guard res == noErr else {
      throw SRAudioError.osStatusError(status: res, description: "SRAUGraph.open()")
    }
  }
  
  open func close() throws {
    let res = AUGraphClose(self.graph)
    guard res == noErr else {
      throw SRAudioError.osStatusError(status: res, description: "SRAUGraph.close()")
    }
  }
  
  open func nodeInfo(_ node: SRAUNode) throws -> SRAudioUnit {
    var audioUnit: AudioUnit? = nil
    let res = AUGraphNodeInfo(self.graph, node.node, nil, &audioUnit)
    guard res == noErr else {
      throw SRAudioError.osStatusError(status: res, description: "SRAUGraph.nodeInfo()")
    }
    
    return SRAudioUnit(audioUnit: audioUnit!)
  }
  
  open func connect(sourceNode: SRAUNode, sourceOutputNumber: UInt32, destNode: SRAUNode, destInputNumber: UInt32) throws {
    let res = AUGraphConnectNodeInput(self.graph, sourceNode.node, sourceOutputNumber, destNode.node, destInputNumber)
    guard res == noErr else {
      throw SRAudioError.osStatusError(status: res, description: "SRAUGraph.connect()")
    }
  }
  
  open func setNodeInputCallback(_ destNode: SRAUNode, destInputNumber: UInt32, callback: AURenderCallbackStruct) throws {
    var callbackRef = callback
    let res = AUGraphSetNodeInputCallback(self.graph, destNode.node, destInputNumber, &callbackRef)
    guard res == noErr else {
      throw SRAudioError.osStatusError(status: res, description: "SRAUGraph.setNodeInputCallback()")
    }
  }
  
  open func addRenderNotify(userData procRefCon: UnsafeMutableRawPointer, callback: @escaping AURenderCallback) throws {
    let res = AUGraphAddRenderNotify(self.graph, callback, procRefCon)
    guard res == noErr else {
      throw SRAudioError.osStatusError(status: res, description: "SRAUGraph.addRenderNotify()")
    }
  }
  
  open func initialize() throws {
    let res = AUGraphInitialize(self.graph)
    guard res == noErr else {
      throw SRAudioError.osStatusError(status: res, description: "SRAUGraph.initialize()")
    }
  }
  
  open func uninitialize() throws {
    let res = AUGraphUninitialize(self.graph)
    guard res == noErr else {
      throw SRAudioError.osStatusError(status: res, description: "SRAUGraph.uninitialize()")
    }
  }
  
  open func update(_ sync: Bool = true) throws {
    let res: OSStatus
    
    if sync {
      res = AUGraphUpdate(self.graph, nil)
    } else {
      var updateValue = DarwinBoolean(false)
      res = AUGraphUpdate(self.graph, &updateValue)
    }
    
    guard res == noErr else {
      throw SRAudioError.osStatusError(status: res, description: "SRAUGraph.update()")
    }
  }
  
  open func start() throws {
    let res = AUGraphStart(self.graph)
    guard res == noErr else {
      throw SRAudioError.osStatusError(status: res, description: "SRAUGraph.start()")
    }
  }
  
  open func stop() throws {
    let res = AUGraphStop(self.graph)
    guard res == noErr else {
      throw SRAudioError.osStatusError(status: res, description: "SRAUGraph.stop()")
    }
  }
  
  open func CAShow() {
    SRAudioCAShow(self.graph)
  }
}
